File: StoryBrew/Services/AudioService.swift
import Foundation
import AVFoundation

/// Reads a story aloud using on-device text-to-speech, giving each character a
/// distinct voice via their `VoiceProfile` (voice + pitch + rate).
///
/// Audio is synthesised on demand each time a story is played, so there are no
/// audio files to store or clean up — replaying is just re-running synthesis.
@MainActor
final class AudioService: NSObject, ObservableObject {
    @Published private(set) var isPlaying = false
    @Published private(set) var isPaused = false
    /// Index into the currently-playing story's `lines`, or nil when stopped.
    @Published private(set) var currentLineIndex: Int?
    /// The story currently loaded into the player.
    @Published private(set) var currentStoryID: UUID?

    private let synthesizer = AVSpeechSynthesizer()
    private var lines: [StoryLine] = []
    private var voiceResolutions: [String: AVSpeechSynthesisVoice?] = [:]
    private var availableEnglishVoices: [AVSpeechSynthesisVoice] = []
    /// Counters so progress tracking is correct even if two lines share text.
    private var startedCount = 0
    private var finishedCount = 0

    override init() {
        super.init()
        synthesizer.delegate = self
        availableEnglishVoices = AVSpeechSynthesisVoice.speechVoices()
            .filter { $0.language.hasPrefix("en") }
    }

    // MARK: - Controls

    func play(_ story: Story) {
        stop()
        configureAudioSession()
        currentStoryID = story.id
        lines = story.lines
        voiceResolutions = resolveVoices(for: story)
        guard !lines.isEmpty else { return }

        startedCount = 0
        finishedCount = 0
        isPlaying = true
        isPaused = false
        for line in lines {
            synthesizer.speak(utterance(for: line, in: story))
        }
        currentLineIndex = 0
    }

    func pause() {
        guard isPlaying, !isPaused else { return }
        synthesizer.pauseSpeaking(at: .word)
        isPaused = true
    }

    func resume() {
        guard isPlaying, isPaused else { return }
        synthesizer.continueSpeaking()
        isPaused = false
    }

    func togglePlayPause(for story: Story) {
        if currentStoryID == story.id, isPlaying {
            isPaused ? resume() : pause()
        } else {
            play(story)
        }
    }

    func stop() {
        if synthesizer.isSpeaking {
            synthesizer.stopSpeaking(at: .immediate)
        }
        isPlaying = false
        isPaused = false
        currentLineIndex = nil
        currentStoryID = nil
        lines = []
    }

    /// Whether this story is the one currently loaded in the player.
    func isCurrent(_ story: Story) -> Bool { currentStoryID == story.id }

    // MARK: - Synthesis helpers

    private func utterance(for line: StoryLine, in story: Story) -> AVSpeechUtterance {
        let profile = story.voiceProfile(for: line.speaker)
        let utterance = AVSpeechUtterance(string: line.text)
        utterance.pitchMultiplier = profile.pitch
        utterance.rate = profile.rate
        utterance.preUtteranceDelay = 0.15
        utterance.postUtteranceDelay = 0.25
        if let voice = voiceResolutions[profile.id] ?? nil {
            utterance.voice = voice
        }
        return utterance
    }

    /// Pick a concrete installed voice for each profile used in the story.
    private func resolveVoices(for story: Story) -> [String: AVSpeechSynthesisVoice?] {
        var profiles = story.characters.map(\.voiceProfile)
        profiles.append(.narrator)

        var result: [String: AVSpeechSynthesisVoice?] = [:]
        for (offset, profile) in profiles.enumerated() {
            result[profile.id] = pickVoice(for: profile, salt: offset)
        }
        return result
    }

    private func pickVoice(for profile: VoiceProfile, salt: Int) -> AVSpeechSynthesisVoice? {
        guard !availableEnglishVoices.isEmpty else {
            return AVSpeechSynthesisVoice(language: "en-US")
        }
        let candidates: [AVSpeechSynthesisVoice]
        switch profile.genderLean {
        case .feminine:
            candidates = availableEnglishVoices.filter { $0.gender == .female }
        case .masculine:
            candidates = availableEnglishVoices.filter { $0.gender == .male }
        case .neutral:
            candidates = availableEnglishVoices
        }
        let pool = candidates.isEmpty ? availableEnglishVoices : candidates
        // Stable pick so the same character keeps the same base voice.
        let index = abs(profile.id.hashValue &+ salt) % pool.count
        return pool[index]
    }

    private func configureAudioSession() {
        #if os(iOS)
        let session = AVAudioSession.sharedInstance()
        try? session.setCategory(.playback, mode: .spokenAudio, options: [])
        try? session.setActive(true)
        #endif
    }
}

extension AudioService: AVSpeechSynthesizerDelegate {
    nonisolated func speechSynthesizer(_ synthesizer: AVSpeechSynthesizer,
                                       didStart utterance: AVSpeechUtterance) {
        Task { @MainActor in
            guard isPlaying else { return }
            currentLineIndex = min(startedCount, max(0, lines.count - 1))
            startedCount += 1
        }
    }

    nonisolated func speechSynthesizer(_ synthesizer: AVSpeechSynthesizer,
                                       didFinish utterance: AVSpeechUtterance) {
        Task { @MainActor in
            guard isPlaying else { return }
            finishedCount += 1
            if finishedCount >= lines.count {
                stop()
            }
        }
    }
}

File: src/services/audioService.ts
import * as Speech from 'expo-speech';
import { Platform } from 'react-native';
import { Audio, InterruptionModeIOS, InterruptionModeAndroid } from 'expo-av';
import { Story, StoryLine, VoiceProfile, NARRATOR_SPEAKER } from '@/models/types';
import { NARRATOR_VOICE } from '@/models/voices';
import { setNowPlaying, setPlaybackState, clearNowPlaying } from './mediaSession';

export interface AudioState {
  currentStoryId: string | null;
  isPlaying: boolean;
  isPaused: boolean;
  currentLineIndex: number | null;
}

/**
 * Reads a story aloud using on-device text-to-speech (expo-speech), giving each
 * character a distinct voice via their VoiceProfile (voice + pitch + rate).
 *
 * Plays lines sequentially: each line speaks in its speaker's voice, and the
 * next line starts when the previous one finishes. Pause is implemented by
 * stopping and remembering the position, then re-speaking from there on resume
 * — this behaves identically on iOS, Android and the web (where native
 * pause/resume support varies).
 *
 * Audio is synthesised on demand each play, so there are no files to store.
 */
export class AudioService {
  private state: AudioState = {
    currentStoryId: null,
    isPlaying: false,
    isPaused: false,
    currentLineIndex: null,
  };

  private listeners = new Set<(s: AudioState) => void>();
  private lines: StoryLine[] = [];
  private voiceByProfileId: Record<string, string | undefined> = {};
  private generation = 0;
  private englishVoiceIds: string[] = [];
  private voicesLoaded = false;
  private audioSessionReady = false;

  // ---- Subscription (for React) ----

  subscribe = (listener: (s: AudioState) => void): (() => void) => {
    this.listeners.add(listener);
    return () => this.listeners.delete(listener);
  };

  getSnapshot = (): AudioState => this.state;

  private setState(patch: Partial<AudioState>) {
    this.state = { ...this.state, ...patch };
    this.listeners.forEach((l) => l(this.state));
  }

  // ---- Controls ----

  isCurrent(storyId: string): boolean {
    return this.state.currentStoryId === storyId;
  }

  async play(story: Story): Promise<void> {
    this.stop();
    await this.configureAudioSession();
    await this.ensureVoicesLoaded();

    this.lines = story.lines;
    this.voiceByProfileId = this.resolveVoices(story);
    if (this.lines.length === 0) return;

    this.generation += 1;
    this.setState({
      currentStoryId: story.id,
      isPlaying: true,
      isPaused: false,
      currentLineIndex: 0,
    });
    // Surface the story to OS / car / Bluetooth media controls (web).
    setNowPlaying(
      { title: story.title, artist: story.seriesTitle ?? 'StoryBrew' },
      {
        onPlay: () => this.resume(story),
        onPause: () => this.pause(),
        onStop: () => this.stop(),
      }
    );
    setPlaybackState('playing');
    this.speakFrom(0, story);
  }

  pause(): void {
    if (!this.state.isPlaying || this.state.isPaused) return;
    this.setState({ isPaused: true });
    this.generation += 1; // invalidate the in-flight line's onDone
    Speech.stop();
    setPlaybackState('paused');
  }

  resume(story: Story): void {
    if (!this.state.isPlaying || !this.state.isPaused) return;
    this.setState({ isPaused: false });
    setPlaybackState('playing');
    this.speakFrom(this.state.currentLineIndex ?? 0, story);
  }

  togglePlayPause(story: Story): void {
    if (this.isCurrent(story.id) && this.state.isPlaying) {
      this.state.isPaused ? this.resume(story) : this.pause();
    } else {
      void this.play(story);
    }
  }

  stop(): void {
    this.generation += 1;
    Speech.stop();
    this.lines = [];
    this.setState({
      currentStoryId: null,
      isPlaying: false,
      isPaused: false,
      currentLineIndex: null,
    });
    clearNowPlaying();
  }

  /**
   * Configure the OS audio session for media playback (once). This is what lets
   * the story route to a connected Bluetooth device, car speaker, or CarPlay,
   * play even when the ringer is on silent, and keep playing with the screen
   * locked / app backgrounded (iOS background audio mode is set in app.json).
   * No-op on web, where output routing is handled by the browser/OS.
   */
  private async configureAudioSession(): Promise<void> {
    if (this.audioSessionReady || Platform.OS === 'web') {
      this.audioSessionReady = true;
      return;
    }
    try {
      await Audio.setAudioModeAsync({
        allowsRecordingIOS: false,
        playsInSilentModeIOS: true,
        staysActiveInBackground: true,
        interruptionModeIOS: InterruptionModeIOS.DoNotMix,
        shouldDuckAndroid: false,
        interruptionModeAndroid: InterruptionModeAndroid.DoNotMix,
        playThroughEarpieceAndroid: false,
      });
    } catch {
      // If the session can't be configured we still attempt playback; it just
      // won't get background/route benefits.
    }
    this.audioSessionReady = true;
  }

  // ---- Internals ----

  private speakFrom(index: number, story: Story): void {
    if (index >= this.lines.length) {
      this.stop();
      return;
    }
    const myGeneration = this.generation;
    const line = this.lines[index];
    const profile = this.profileFor(story, line.speaker);
    this.setState({ currentLineIndex: index });

    Speech.speak(line.text, {
      voice: this.voiceByProfileId[profile.id],
      pitch: profile.pitch,
      rate: profile.rate,
      language: 'en-US',
      onDone: () => {
        // Ignore stale callbacks from a stopped/paused/restarted session.
        if (myGeneration !== this.generation || this.state.isPaused) return;
        this.speakFrom(index + 1, story);
      },
    });
  }

  private profileFor(story: Story, speaker: string): VoiceProfile {
    const match = story.characters.find(
      (c) => c.name.toLowerCase() === speaker.toLowerCase()
    );
    return match ? match.voiceProfile : NARRATOR_VOICE;
  }

  /** Assign a distinct installed voice to each profile used in the story. */
  private resolveVoices(story: Story): Record<string, string | undefined> {
    const profiles: VoiceProfile[] = [
      ...story.characters.map((c) => c.voiceProfile),
      NARRATOR_VOICE,
    ];
    const result: Record<string, string | undefined> = {};
    profiles.forEach((profile, i) => {
      result[profile.id] =
        this.englishVoiceIds.length > 0
          ? this.englishVoiceIds[i % this.englishVoiceIds.length]
          : undefined;
    });
    return result;
  }

  private async ensureVoicesLoaded(): Promise<void> {
    if (this.voicesLoaded) return;
    try {
      const voices = await Speech.getAvailableVoicesAsync();
      this.englishVoiceIds = voices
        .filter((v) => v.language?.toLowerCase().startsWith('en'))
        .map((v) => v.identifier);
    } catch {
      this.englishVoiceIds = [];
    }
    this.voicesLoaded = true;
  }
}
